internal/bufferedwrites/upload_handler.go

```diff
@@ -96,9 +103,16 @@ func (uh *UploadHandler) uploader() {
 	for currBlock := range uh.uploadCh {
 		_, err := io.Copy(uh.writer, currBlock.Reader())
 		if err != nil {
-			logger.Errorf("upload failed: error in io.Copy: %v", err)
-			uh.wg.Done()
-			// TODO: handle failure scenario: finalize the upload and trigger edit flow.
+			logger.Errorf("buffered write upload failed for object %s: error in io.Copy: %v", uh.objectName, err)
+			// Close the channel to signal upload failure.
+			close(uh.signalUploadFailure)
+
+			// Close the writer to finalize the object creation on GCS.
+			if closeErr := uh.writer.Close(); closeErr != nil {
+				logger.Errorf("Error in finalizing object %s: %v", uh.objectName, closeErr)
+				return
+			}
+			return
 		}
 		uh.wg.Done()
```

Comment:
you are not marking it as done for failure case. what happens if finalize is triggered after failure?